src/build/env.py

```diff
@@ -141,27 +188,67 @@ def __enter__(self):  # type: () -> IsolatedEnvironment
+        '''
+        self._place_path_relative(sysconfig.get_path('include'))
+        self._place_path_relative(sysconfig.get_path('platinclude'))
+        self._place_path_relative(sysconfig.get_config_var('LIBPL'))
+
+        '''
+        We use PYTHONHOME to relocate the Python installation to our environment,
+        but turns out this can't be blindely relied upon, *sigh*. If a pyvenv.cfg
+        file is present on the same directory than the binary or one level above,
+        data, platlib, platstdlib, purelib and scripts will be overritten. So we
+        need to make sure that doesn't happen, hence we symlink the Python
+        executable onto a path we control and use that binary for the enrionment.
+
+        Unfortunately, this forces us to override sys.executable. There is no
+        other way of doing it.
+        '''
+        executable_dir = 'executable'
+        executable_path = os.path.join(self.path, executable_dir)
+        self._place_file(sys.executable, os.path.join(executable_dir, 'python'))
+        exe_path.insert(0, executable_path)
+
+        self._executable = sys.executable
+        sys.executable = os.path.join(executable_path, 'python')
+
+        # Replace PATH with our value for the environment
+        self._replace_env('PATH', os.pathsep.join(sorted(
+            set(exe_path), key=exe_path.index
+        )))  # sorted is required for Python 2 and 3.5
+
+        # Inject the missing import paths via PYTHONPATH
         self._replace_env('PYTHONPATH', os.pathsep.join(sys_path))
+
+        # Point the Python interpreter to our environment
         self._replace_env('PYTHONHOME', self.path)
-        self._pop_env('PIP_REQUIRE_VIRTUALENV')
 
-        self._symlink_relative(sysconfig.get_path('include'))
-        self._symlink_relative(sysconfig.get_path('platinclude'))
-        self._symlink_relative(sysconfig.get_config_var('LIBPL'))
+        # Remove environment variables that intrefer with our use of pip
+        self._pop_env('PIP_REQUIRE_VIRTUALENV')
 
         return self
 
     def __exit__(self, typ, value, traceback):
         # type: (Optional[Type[BaseException]], Optional[BaseException], Optional[types.TracebackType]) -> None
+        '''
+        Restores the everything to the original state
+        '''
         if self.path and os.path.isdir(self.path):
             shutil.rmtree(self.path)
 
         self._restore_env()
+        sys.executable = self._executable
 
     def install(self, requirements):  # type: (Iterable[str]) -> None
         '''
-        Installs the specified requirements on the environment
+        Installs the specified PEP 508 requirements on the environment
+
+        Passing non PEP 508 strings will result in undefined behavior, you
+        *should not* rely on it. It is merely an implementation detail, it may
+        change any time without warning.
         '''
         if not requirements:
             return
```

Comment:
I just realized that you are calling:

```
subprocess.check_call([sys.executable, '-m', 'ensurepip'], cwd=self.path)
```

This will install pip and setuptools. IMHO this is bad because now setuptools is by default provisioned and not specifying it within your build requires will silently pass. We should ensure only pip is installed in there 🤔 . Perhaps we should uninstall if not specified within the build-requires list. On handle the easy to make error of forgetting setuptools as build-require but build still passes.

Reply:
You are right, I hadn't noticed before. We should be able to just delete `setuptools*` from `site-packages`.

Reply:
> We should be able to just delete `setuptools*` from `site-packages`.

I don't think that's wise, you'd be relying there on implementation details of what pip uninstall does. Probably better to just use pip uninstall directly.

Reply:
WTF? Why does `ensurepip` install `setuptools`?

Reply:
Because in a source only non PEP-518 world ``pip install`` would break without a local setuptools.. Is my guess.